src/DynamoCoreWpf/Views/Core/NoteView.xaml.cs

```diff
@@ -152,5 +168,268 @@ private void PrepareZIndex()
+        {
+            // In order to get the caretIndex relative to the caret line:
+            // Get the text untill the caretIndex
+            // Split the text into lines
+            // The caret index relative to the caret line will be equal to
+            // the last line's length
+            var textUntillCaretIndex = text.Substring(0, caretIndex);
+            var textInLines = BreakTextIntoLines(textUntillCaretIndex);
+            var caretRelativeToLine = textInLines.Last().Length;
+            return caretRelativeToLine;
+        }
+        
+        private int GetLineNumberAtCaretsIndex(string text, int caretIndex)
+        {
+            var textBeforeCaret = text.Substring(0, caretIndex);
+            var textInLines = BreakTextIntoLines(textBeforeCaret);
+            return textInLines.Length-1;
+        }
+        private string GetLineTextAtCaretsIndex(string text, int caretIndex)
+        {
+            var textInLines = BreakTextIntoLines(text);
+            var caretsLineIndex = GetLineNumberAtCaretsIndex(text, caretIndex);
+            return textInLines[caretsLineIndex];
+        }
+
+        private string ReplaceLineOfText(string text, int lineNumber, string newLine)
+        {
+            var textInLines = BreakTextIntoLines(text);
+            textInLines[lineNumber] = newLine;
+            return String.Join("\n", textInLines);
+        }
+
+        private string UpdateBulletAccordingToIndentation(string text, bool reverse = false)
+        {
+            for (int i = 0; i < BULLETS_CHARS.Length; i++)
+            {
+                int nextBulletIndex = (i + 1) % BULLETS_CHARS.Length;
+                if (reverse)
+                {
+                    nextBulletIndex = (i + BULLETS_CHARS.Length - 1) % BULLETS_CHARS.Length;
+                }
+
+                var newText = text.Replace(BULLETS_CHARS[i], BULLETS_CHARS[nextBulletIndex]);
+                if (!String.Equals(text, newText))
+                {
+                    text = newText;
+                    break;
+                }
+            }
+            return text;
+        }
+
+        /// <summary>
+        /// Following suggestions from stackoverflow,
+        /// A reliable method for breaking text into lines
+        /// using Regex is used.
+        /// https://stackoverflow.com/questions/1508203/best-way-to-split-string-into-lines
+        /// </summary>
+        /// <param name="text"> text to break into lines</param>
+        /// <returns> text lines </returns>
+        private string[] BreakTextIntoLines(string text)
+        {
+            return Regex.Split(text, "\r\n|\r|\n");
+        }
+
+        private string TabToSpaceConversion(string text)
+        {
+            return text.Replace("\t", new String(' ', TAB_SPACING_SIZE));
+        }
+
+        private string SpaceToTabConversion(string text)
+        {
+            return text.Replace(new String(' ', TAB_SPACING_SIZE), "\t");
+        }
```

Comment:
I feel a few of these fall into the category of Utility functions which could be shared.. Would you create `StringUtils.cs` under `DynamoUtilities` project and move some of the sprint manipulation functions over?

Reply:
ok, @QilongTang, I just added the functions that are applicable to all strings to the `StringUtils.cs` class which are:
`ReplaceLineOfText` , `IsStringSpacesWithTabs` , `BreakTextIntoLines` , `TabToSpaceConversion` , `SpaceToTabConversion`
Ive kept the ones that are exclusively related to bullet point support in `NoteView.xaml.cs` which are:
`BulletDashHandler` , `BulletTabHandler` , `BulletEnterHandler` , `IsCaretRightAfterBullet` , `IsCaretBeforeABullet` , `GetCaretIndexRelativeToLine` , `GetLineNumberAtCaretsIndex` , `GetLineTextAtCaretsIndex` , and `UpdateBulletAccordingToIndentation`
let me know if this is ok

Reply:
addressed in 
28b73d9